pyo3-macros-backend/src/intopydict.rs

```diff
@@ -0,0 +1,366 @@
+use quote::{quote, TokenStreamExt};
+use std::{collections::HashMap, ops::AddAssign};
+
+use proc_macro2::{Span, TokenStream};
+use syn::{
+    parse::{Parse, ParseStream},
+    DeriveInput, Error, Generics,
+};
+
+const COL_NAMES: [&str; 8] = [
+    "BTreeSet",
+    "BinaryHeap",
+    "Vec",
+    "HashSet",
+    "LinkedList",
+    "VecDeque",
+    "BTreeMap",
+    "HashMap",
+];
```

Comment:
For autoref is there a particular crate that the rest of pyo3 uses/you reccomend?

Reply:
Would this apply to collections as well (ie Vecs, Sets etc)? These would get stored as a PyObject instead of a PyList?

Reply:
And how do you recommend I use autoref specialization? I don't have much experience with it. Is there a crate which makes it easier?

Reply:
A separate crate usually doesn't help because abstracting the pattern way is often more complex than just applying it.

In this case, the general technique is described at https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md and you can an example of its application in PyO3's existing macros by looking for usages of the type `PyClassImplCollector` defined at 

https://github.com/PyO3/pyo3/blob/ae982b8ad0fff71d4d2fdbc2f7055500a6fb450f/src/impl_/pyclass.rs#L109

and for example its relation to the trait `PyClassNewTextSignature` defined at

https://github.com/PyO3/pyo3/blob/ae982b8ad0fff71d4d2fdbc2f7055500a6fb450f/src/impl_/pyclass.rs#L995

Reply:
Should I add the specialization feature as a crate attribute?

Reply:
No; the specialization feature is incomplete and unsound. Autoref specialization is a technique which works on stable Rust to get a specialization-like result for some limited cases.

Reply:
There seems to be some issues with reference moving when I try to do autoref specialization. Sorry it took me a while for this, I was finishing up an internship